File: app/src/main/java/com/ue/colorful/feature/test/ColorVisionTestFragment.java
package com.ue.colorful.feature.test;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.ue.colorful.R;

public class ColorVisionTestFragment extends Fragment {
    private static final String ARG_NUM = "arg_num";

    private String[] testAnswer;
    private static int[] testImg = new int[]{
            R.raw.p01, R.raw.p02, R.raw.p03, R.raw.p04, R.raw.p05,
            R.raw.p06, R.raw.p07, R.raw.p08, R.raw.p09, R.raw.p10,
            R.raw.p11, R.raw.p12, R.raw.p13, R.raw.p14, R.raw.p15, R.raw.p16,
            R.raw.p17, R.raw.p18, R.raw.p19, R.raw.p20, R.raw.p21, R.raw.p22,
            R.raw.p23, R.raw.p24,
    };

    int pageNum;

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment testImg.
     */
    public static ColorVisionTestFragment newInstance(int num) {
        ColorVisionTestFragment fragment = new ColorVisionTestFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_NUM, num);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            pageNum = getArguments().getInt(ARG_NUM);
        }
        testAnswer = getResources().getStringArray(R.array.colorVisionAns);
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_test_img, container, false);
        ImageView img = (ImageView) view.findViewById(R.id.testImg);
        img.setImageResource(testImg[pageNum]);
        TextView text = view.findViewById(R.id.answer);
        text.setText(testAnswer[pageNum]);
        return view;
    }
}

File: app/src/main/java/com/ue/colorful/feature/test/ColorVisionTestFragment.kt
package com.ue.colorful.feature.test

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.ViewPager
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.squareup.picasso.Picasso
import com.ue.colorful.R
import com.ue.colorful.feature.main.BaseFragment
import kotlinx.android.synthetic.main.fragment_color_vision_test.view.*
import kotlinx.android.synthetic.main.fragment_test_img.view.*

/**
 * Created by hawk on 2017/12/14.
 */
class ColorVisionTestFragment : BaseFragment(R.layout.fragment_color_vision_test, 0) {
    override fun initViews() {
        val testAnswer = resources.getStringArray(R.array.colorVisionAns)
        val size = testAnswer.size
        val states = SparseArray<Boolean>()

        val testImgResTa = resources.obtainTypedArray(R.array.colorVisionImgs)
        val testImgRes = Array(size, { i -> testImgResTa.getResourceId(i, 0) })
        testImgResTa.recycle()

        //绑定自定义适配器
        rootView.vpTestPager.adapter = object : FragmentStatePagerAdapter(childFragmentManager) {
            override fun getCount(): Int {
                return size
            }

            override fun getItem(position: Int): Fragment {
                val fragment = CVTPagerFragment.newInstance(position, testAnswer[position], testImgRes[position], states[position, false])
                fragment.setAnalyseListener(object : View.OnClickListener {
                    override fun onClick(v: View?) {
                        states.put(position, true)
                    }
                })
                return fragment
            }
        }

        rootView.tvTags.text = "${(rootView.vpTestPager.currentItem + 1)}/$size"

        rootView.vpTestPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
            }

            override fun onPageSelected(position: Int) {
                rootView.tvTags.text = "${(rootView.vpTestPager.currentItem + 1)}/$size"
            }

            override fun onPageScrollStateChanged(state: Int) {
            }
        })
    }

    class CVTPagerFragment : Fragment() {
        private var pageNum: Int = 0
        private var testAns: String = ""
        private var testImg: Int = 0
        private var isShow: Boolean = false
        private var onAnalyseListener: View.OnClickListener? = null

        fun setAnalyseListener(onAnalyseListener: View.OnClickListener) {
            this.onAnalyseListener = onAnalyseListener
        }

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)

            pageNum = arguments.getInt(ARG_NUM)
            testAns = arguments.getString(ARG_TEST_ANS)
            testImg = arguments.getInt(ARG_TEST_IMG)
            isShow = arguments.getBoolean(ARG_IS_VISIBLE)
        }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            val view = inflater.inflate(R.layout.fragment_test_img, container, false)

            Picasso.with(activity)
                    .load(testImg)
                    .into(view.testImg)

            if (isShow) {
                view.tvTips.visibility = View.GONE
                view.vgAnalysePanel.isSelected = true
                (view.vsAnswer.inflate() as TextView).text = testAns
            } else {
                view.tvTips.setOnClickListener { _ ->
                    view.tvTips.visibility = View.GONE
                    view.vgAnalysePanel.isSelected = true
                    (view.vsAnswer.inflate() as TextView).text = testAns
                    onAnalyseListener?.onClick(null)
                }
            }

            return view
        }

        companion object {
            private val ARG_NUM = "arg_num"
            private val ARG_TEST_ANS = "arg_test_ans"
            private val ARG_TEST_IMG = "arg_test_img"
            private val ARG_IS_VISIBLE = "arg_is_visible"

            fun newInstance(num: Int, testAns: String, testImg: Int, isVisible: Boolean): CVTPagerFragment {
                val fragment = CVTPagerFragment()
                val args = Bundle()
                args.putInt(ARG_NUM, num)
                args.putString(ARG_TEST_ANS, testAns)
                args.putInt(ARG_TEST_IMG, testImg)
                args.putBoolean(ARG_IS_VISIBLE, isVisible)
                fragment.arguments = args
                return fragment
            }
        }
    }

}